#Tp53 Science Ebert GSE131592 MOLM13 cell line
#Data preprocessing and QC metrics
#Author:Alba Rodriguez-Meira
#Date: November 2020
#Modified last: 13th March 2021
####################################################

library(ggplot2)
library(reshape2)

#Import counts tables from featureCounts - all reads from all cells

counts<-read.table("input/counts_GSE131592.txt",header = T,sep="\t") #Read table generated by FeatureCounts
metadata<- read.table("input/metadata_GSE131592.txt",header = T,sep="\t")
genes<-read.table("input/genes_withERCC_uniqueids.tsv",header = F,sep = "\t")

#Format counts table
gene.ids <- counts$Geneid
gene.lengths <- counts$Length
counts <- as.matrix(counts[,-c(1:6)])
colnames(counts)<-gsub(x = colnames(counts), pattern = "(.Aligned.sortedByCoord.out.bam)", replacement = "")
rownames(counts) <- genes$V4

#Format metadata table
rownames(metadata)<-metadata$Sample_ID
counts<-counts[,metadata$Sample_ID]

#Create data frame with all information from sequencing statistics
colData = data.frame(cell.id=colnames(counts), lib.size=colSums(counts))
rowData = data.frame(gene.id = rownames(counts),type=ifelse(grepl("ERCC-",rownames(counts)),"ERCC","gene"))
colData$ercc.count <- colSums(counts[rowData$type == "ERCC",])

#################################################################################

### Perform counts normalization

normalize_counts <- function(your.matrix,scale.factor = 1e4,use.scaled.factor=T){
  
  umi.dat<-your.matrix
  if(use.scaled.factor==T){
    totalUMI.lib<-Matrix::colSums(umi.dat)
    normalized.umi<-(t(t(umi.dat)/totalUMI.lib))*scale.factor
    return(normalized.umi)
  }else{
    totalUMI.lib<-Matrix::colSums(umi.dat)
    normalized.umi<-(t(t(umi.dat)/totalUMI.lib))*round(mean(totalUMI.lib))
    return(normalized.umi)
  }
  
}

#mean: 14240792 (14 M)
counts.normalized<-normalize_counts(counts,use.scaled.factor=F)

#Calculate number of genes with counts>1 
counts_greater1 <- data.frame(cell.id= colnames(counts.normalized),genes.detected=as.numeric(apply(counts.normalized >= 1,2, sum)))

#Add information about the number of genes to colData
counts_greater1$cell.id<-as.character(counts_greater1$cell.id)
colData <- data.frame(counts_greater1[match(colData$cell.id,counts_greater1$cell.id),], colData)

#Merge colData with metadata
colData <- data.frame(metadata[match(colData$cell.id,as.character(metadata$Sample_ID)),], colData)
write.table(colData,file="colData_Ebert.txt",sep="\t",row.names = T)
write.table(counts.normalized,file="counts.normalized.Ebert.txt",sep="\t",row.names = T)

#########################################################################################################

# colData$status<-"mutant"
# colData$status[colData$Genotype=="WT"]<-"aWT"
# 
# colData$CEBPA<-counts.normalized["CEBPA",]
# colData$GATA1<-counts.normalized["GATA1",]
# colData$KLF1<-counts.normalized["KLF1",]
# 
# ggplot(colData, aes(x=status, y=CEBPA,fill=status)) + geom_boxplot()+
#   theme_classic()+
#   scale_fill_manual(values = c("grey","red"))+
#   theme(axis.text.y = element_text(size=20),legend.position = "none",
#         axis.text.x=element_blank(),axis.ticks.x = element_blank(),
#         axis.title.x = element_blank())
# 
# ggsave('figures/CEBPA_per.genotype.png',height=5,width=3)
# 
# t.test(colData[colData$status=="mutant",c("CEBPA")],colData[colData$status=="aWT",c("CEBPA")])
# #p-value = 0.7419
# fc=11.12008/12.71889 #0.8742964
# 
# ggplot(colData, aes(x=status, y=GATA1,fill=status)) + geom_boxplot()+
#   theme_classic()+
#   scale_fill_manual(values = c("grey","red"))+
#   theme(axis.text.y = element_text(size=20),legend.position = "none",
#         axis.text.x=element_blank(),axis.ticks.x = element_blank(),
#         axis.title.x = element_blank())
# 
# ggsave('figures/GATA1_per.genotype.png',height=5,width=3)
# 
# t.test(colData[colData$status=="mutant",c("GATA1")],colData[colData$status=="aWT",c("GATA1")])
# #p-value = 0.001834
# 
# ggplot(colData, aes(x=status, y=KLF1,fill=status)) + geom_boxplot()+
#   theme_classic()+
#   scale_fill_manual(values = c("grey","red"))+
#   theme(axis.text.y = element_text(size=20),legend.position = "none",
#         axis.text.x=element_blank(),axis.ticks.x = element_blank(),
#         axis.title.x = element_blank())
# 
# ggsave('figures/KLF1_per.genotype.png',height=5,width=3)
# 
# t.test(colData[colData$status=="mutant",c("KLF1")],colData[colData$status=="aWT",c("KLF1")])
# #p-value = 0.4237
# fc=0.8753549/0.4296846 #2.037203
# 
# ######
# erythroid_genes<-c("SLC4A1","BCL11A","CA1","EPOR","HBB","GYPA","TFRC")
# myeloid_genes<-c("CD33","MPO","ITGAM","IRF8")
# 
# counts.normalized.erythroid<-counts.normalized[erythroid_genes,]
# counts.normalized.myeloid<-counts.normalized[myeloid_genes,]
# ######
# colData.mutant<-subset(colData,status=="mutant")
# colData.WT<-subset(colData,status=="aWT")
# 
# counts.normalized.erythroid.mutant<-counts.normalized.erythroid[,colData.mutant$Sample_ID]
# means.erythroid.mutant<-as.data.frame(rowMeans(counts.normalized.erythroid.mutant))
# colnames(means.erythroid.mutant)<-"mutant"
# 
# counts.normalized.erythroid.WT<-counts.normalized.erythroid[,colData.WT$Sample_ID]
# means.erythroid.WT<-as.data.frame(rowMeans(counts.normalized.erythroid.WT))
# colnames(means.erythroid.WT)<-"WT"
# 
# means.erythroid<-cbind(means.erythroid.WT,means.erythroid.mutant)
# ######
# 
# library(pheatmap)
# pheatmap(as.matrix(means.erythroid),scale="row",cluster_rows=FALSE,cluster_cols=FALSE,
#          filename="figures/erythroid.genes.counts.png",height=5,width=3)
# ######
# 
# counts.normalized.myeloid.mutant<-counts.normalized.myeloid[,colData.mutant$Sample_ID]
# means.myeloid.mutant<-as.data.frame(rowMeans(counts.normalized.myeloid.mutant))
# colnames(means.myeloid.mutant)<-"mutant"
# 
# counts.normalized.myeloid.WT<-counts.normalized.myeloid[,colData.WT$Sample_ID]
# means.myeloid.WT<-as.data.frame(rowMeans(counts.normalized.myeloid.WT))
# colnames(means.myeloid.WT)<-"WT"
# 
# means.myeloid<-cbind(means.myeloid.WT,means.myeloid.mutant)
# ######
# 
# pheatmap(as.matrix(means.myeloid),scale="row",cluster_rows=FALSE,cluster_cols=FALSE,
#          filename="figures/myeloid.genes.counts.png",height=4,width=3)
# ######